ext/standard/basic_functions.c

```diff
@@ -1131,18 +1131,34 @@ PHP_FUNCTION(flush)
 /* {{{ Delay for a given number of seconds */
 PHP_FUNCTION(sleep)
 {
-	zend_long num;
+	zval *num;
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_LONG(num)
+		Z_PARAM_NUMBER(num)
 	ZEND_PARSE_PARAMETERS_END();
-
-	if (num < 0) {
-		zend_argument_value_error(1, "must be greater than or equal to 0");
-		RETURN_THROWS();
-	}
-
-	RETURN_LONG(php_sleep((unsigned int)num));
+    if (Z_TYPE_P(num) == IS_DOUBLE) {
+		const double seconds = Z_DVAL_P(num);
+		if (UNEXPECTED(seconds < 0)) {
+			zend_argument_value_error(1, "must be greater than or equal to 0");
+			RETURN_THROWS();
+		}
+#ifdef HAVE_USLEEP
+		const unsigned int fraction_microseconds = (unsigned int)((seconds - (unsigned int)seconds) * 1000000);
+		if(fraction_microseconds > 0) {
+			usleep(fraction_microseconds);
+		}
+#endif
+		RETURN_LONG(php_sleep((unsigned int)seconds));
```

Comment:
so this basically means that it will return 0 if you try to wait 0.1, right? I think this change does not work too well with the return value. You could potentially just return what's requested if usleep returns 0 but you would need to change return type as well.

Reply:
well the userland usleep() function also ignores errors and return nothing; following the usleep() approach for fractional seconds was the easiest solution.. 

If we return "the number of seconds left to sleep" for usleep errors, we would have to change the return value of sleep from int to int|float too 🤔 

can be done, but don't think it's worth the hassle?

Reply:
ok yeah it will actually return 0 most of the time (maybe always) anyway.  it's really edge case.

Btw does it make any sense to call sleep if the number of uint seconds is 0?

Reply:
> it's really edge case

yeah i suggest we ignore it (unless someone feels strongly about it)

>Btw does it make any sense to call sleep if the number of uint seconds is 0?

Yes it does! this is the closest thing PHP userland has to [pthread_yield()](https://man7.org/linux/man-pages/man3/pthread_yield.3.html)
- a `for(;;){}` loop will use 100% of a cpu core, a `for(;;){sleep(0);}` wil usuallyl use <10%, it will constantly be yielding like pthread_yield()